### PR TITLE
do not save plain-text ETag when encryption is requested

### PR DESCRIPTION
**Pull request by harshavardhana**
## Description
do not save plain-text ETag when encryption is requested

## Motivation and Context
fixes an issue under bucket replication could cause
ETags for replicated SSE-S3 single-part PUT objects,
to fail as we would attempt a decryption while listing,
or stat() operation.

## How to test this PR?
```bash
set -x

export CI=1

make || exit -1

killall -9 minio

rm -rf /tmp/xl/
mkdir -p /tmp/xl/1/ /tmp/xl/2/

export MINIO_KMS_SECRET_KEY="my-minio-key:OSMM+vkKUTCvQs9YL/CVMIMt43HFhkUpqJxTmGl6rYw="

NODES=4

args1=()
args2=()
for i in $(seq 1 $NODES); do
        args1+=("http://localhost:$((9000 + $i))/tmp/xl/1/$i ")
        args2+=("http://localhost:$((9100 + $i))/tmp/xl/2/$i ")
done

for i in $(seq 1 $NODES); do
        ./minio server --address "127.0.0.1:$((9000 + $i))" ${args1[@]} & # | tee /tmp/minio/node.$i &
        ./minio server --address "127.0.0.1:$((9100 + $i))" ${args2[@]} & # | tee /tmp/minio/node.$i &
done

sleep 10

./mc alias set myminio1 http://localhost:9001 minioadmin minioadmin
./mc alias set myminio2 http://localhost:9101 minioadmin minioadmin

sleep 1

./mc mb myminio1/testbucket/ --with-lock
./mc mb myminio2/testbucket/ --with-lock
./mc encrypt set sse-s3 my-minio-key myminio1/testbucket/
./mc encrypt set sse-s3 my-minio-key myminio2/testbucket/

./mc replicate add myminio1/testbucket --remote-bucket http://minioadmin:minioadmin@localhost:9101/testbucket --priority 1

sleep 1

./mc cp internal.tar myminio1/testbucket/dir/1.tar
./mc cp internal.tar myminio2/testbucket/dir/2.tar

sleep 1

./mc ls myminio1/testbucket/dir/

./mc ls myminio2/testbucket/dir/
```

## Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Optimization (provides speedup with no functional changes)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [ ] Fixes a regression (If yes, please add `commit-id` or `PR #` here)
- [x] Unit tests added/updated
- [ ] Internal documentation updated
- [ ] Create a documentation update request [here](https://github.com/minio/docs/issues/new?label=doc-change,title=Doc+Updated+Needed+For+PR+github.com%2fminio%2fminio%2fpull%2fNNNNN)

### CODE REVIEW

File: cmd/erasure-object.go [hunk 1]
```diff
@@ -1263,7 +1264,8 @@ func (er erasureObjects) putObject(ctx context.Context, bucket string, object st
 	}
 
 	userDefined["etag"] = r.MD5CurrentHexString()
-	if opts.PreserveETag != "" {
+	_, ok := crypto.IsEncrypted(userDefined)
```
Comment (vadmeste) [inline cmd/erasure-object.go]: I think S3.IsEncrypted() is better.

We can keep ETags for non encrypted, SSE-KMS, multipart but not SSE-S3

Comment (harshavardhana) [reply cmd/erasure-object.go]: multipart will never come here @vadmeste for preserveETag and in-case of SSE-KMS also we must save the longer encrypted version.

That's why this code says that it must be plain-text only to be able to preserve a custom etag remotely.

**Review by poornas — CHANGES_REQUESTED** [review]
https://github.com/minio/minio/blob/master/cmd/object-handlers-common.go#L193 
etagMatch should be considered true if it is sse-s3 and etag is empty to avoid re-replicating if there are timing issues

File: cmd/erasure-object.go [hunk 2]
```diff
@@ -1263,8 +1264,16 @@ func (er erasureObjects) putObject(ctx context.Context, bucket string, object st
 	}
 
 	userDefined["etag"] = r.MD5CurrentHexString()
+	kind, _ := crypto.IsEncrypted(userDefined)
 	if opts.PreserveETag != "" {
-		userDefined["etag"] = opts.PreserveETag
+		if !opts.ReplicationRequest {
+			userDefined["etag"] = opts.PreserveETag
```
Comment (vadmeste) [inline cmd/erasure-object.go:1270]: @poornas why would we want to preserve ETag in a PutObject request if this is not a replication request ? is this just a defensive code ?

Comment (harshavardhana) [reply cmd/erasure-object.go:1270]: Decommissioning, rebalancing forgot? @vadmeste

Comment (vadmeste) [reply cmd/erasure-object.go:1270]: oh right

**Review by vadmeste — APPROVED** [review]
LGTM

**Review by poornas — APPROVED** [review]

**Pull request merged**